File: lib/claude_inbox/store/sections.rb
# frozen_string_literal: true

module ClaudeInbox
  class Store
    # The Rows of one poll, grouped and ordered. It also knows where the
    # cursor may land, since a folded section stands in for its rows and
    # only Rows with a key take a selection.
    Sections = Struct.new(:pinned, :needs_you, :active, :snoozed, :settled) do
      def each_section = SECTIONS.each { |k| yield k, self[k] }

      def all = SECTIONS.flat_map { |k| self[k] }

      def row(key) = all.find { |r| r.key == key }

      # The rows the `/` filter keeps, in the same sections; the whole thing
      # untouched when there is no filter.
      def matching(query)
        return self if query.nil? || query.empty?
        self.class.new(**to_h.transform_values { |rows| rows.select { |r| r.matches?(query) } })
      end

      # The section a key lives in. A fold's name answers itself.
      def section_of(key)
        return key if FOLDABLE_SECTIONS.include?(key)
        each_section { |name, rows| return name if rows.any? { |r| r.key == key } }
        nil
      end

      # Everything the cursor can land on, top to bottom.
      def selectable_keys(expanded)
        stops(expanded).flat_map { |name, rows| rows ? rows.map(&:key) : [name] }
      end

      # The first stop in each section that has one, as `[name, row]`; the
      # row is nil where the fold itself is the stop.
      def heads(expanded)
        stops(expanded).filter_map { |name, rows| [name, rows&.first] if rows.nil? || rows.any? }
      end

      private

      # Per section, the selectable rows, or nil when the section is folded
      # and the cursor stops on the fold instead. An empty fold is left out:
      # there is nothing under it to open.
      def stops(expanded)
        SECTIONS.filter_map do |name|
          rows = self[name]
          if !Store.folded?(name, expanded) then [name, rows.select(&:selectable?)]
          elsif rows.any? then [name, nil]
          end
        end
      end
    end
  end
end

File: ui/sections.go
package main

import (
	"fmt"
	"strings"
)

var sectionOrder = []string{"pinned", "needs_you", "active", "snoozed", "settled"}

var sectionTitles = map[string]string{
	"pinned": "PINNED", "needs_you": "NEEDS YOU", "active": "ACTIVE", "snoozed": "SNOOZED", "settled": "SETTLED",
}

func foldable(name string) bool { return name == "snoozed" || name == "settled" }

type Sections struct {
	list []Section
}

func (s Sections) Get(name string) []Row {
	for _, sec := range s.list {
		if sec.Name == name {
			return sec.Rows
		}
	}
	return nil
}

func (s Sections) All() []Row {
	var out []Row
	for _, name := range sectionOrder {
		out = append(out, s.Get(name)...)
	}
	return out
}

func (s Sections) Empty() bool { return len(s.All()) == 0 }

func (s Sections) Row(key string) *Row {
	for _, sec := range s.list {
		for i := range sec.Rows {
			if sec.Rows[i].Key == key {
				return &sec.Rows[i]
			}
		}
	}
	return nil
}

// Kept in step with Row#matches? in the Ruby.
func (s Sections) Matching(query string) Sections {
	if query == "" {
		return s
	}
	q := strings.ToLower(query)
	out := Sections{}
	for _, sec := range s.list {
		kept := Section{Name: sec.Name}
		for _, r := range sec.Rows {
			if strings.Contains(strings.ToLower(r.Label), q) || strings.Contains(strings.ToLower(r.Cwd), q) {
				kept.Rows = append(kept.Rows, r)
			}
		}
		out.list = append(out.list, kept)
	}
	return out
}

func (s Sections) Folded(name string, expanded map[string]bool) bool {
	return foldable(name) && !expanded[name]
}

func (s Sections) SectionOf(key string) string {
	if foldable(key) {
		return key
	}
	for _, sec := range s.list {
		for _, r := range sec.Rows {
			if r.Key == key {
				return sec.Name
			}
		}
	}
	return ""
}

type stop struct {
	name string
	rows []Row
	fold bool
}

func (s Sections) stops(expanded map[string]bool) []stop {
	var out []stop
	for _, name := range sectionOrder {
		rows := s.Get(name)
		if !s.Folded(name, expanded) {
			var sel []Row
			for _, r := range rows {
				if r.Selectable {
					sel = append(sel, r)
				}
			}
			out = append(out, stop{name: name, rows: sel})
		} else if len(rows) > 0 {
			out = append(out, stop{name: name, fold: true})
		}
	}
	return out
}

func (s Sections) SelectableKeys(expanded map[string]bool) []string {
	var keys []string
	for _, st := range s.stops(expanded) {
		if st.fold {
			keys = append(keys, st.name)
			continue
		}
		for _, r := range st.rows {
			keys = append(keys, r.Key)
		}
	}
	return keys
}

func (s Sections) Heads(expanded map[string]bool) [][2]string {
	var out [][2]string
	for _, st := range s.stops(expanded) {
		if st.fold {
			out = append(out, [2]string{st.name, st.name})
		} else if len(st.rows) > 0 {
			out = append(out, [2]string{st.name, st.rows[0].Key})
		}
	}
	return out
}

func age(seconds int64) string {
	switch {
	case seconds <= 0:
		return "0s"
	case seconds < 60:
		return fmt.Sprintf("%ds", seconds)
	case seconds < 3600:
		return fmt.Sprintf("%dm", seconds/60)
	case seconds < 86400:
		return fmt.Sprintf("%dh", seconds/3600)
	}
	return fmt.Sprintf("%dd", seconds/86400)
}

func plural(word string, n int) string {
	if n == 1 {
		return word
	}
	return word + "s"
}
